### PR TITLE
Add a physically based metal rough material

### PR DESCRIPTION
**Pull request by nyalldawson**
Adds a new material choice for a physically based metal/roughness material. Options are available for setting the material base color, metalness and roughness.

Internally this uses a clone of Qt's QMetalRoughMaterial class. We use a copy of the Qt class instead of relying on Qt's implementation as longer-term improvements (such as data defined base color) will require a re-implementation anyway. By using our own material we will avoid having two different code paths for the data-defined/ non-data defined scenarios.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk

File: src/ui/3d/metalroughmaterialwidget.ui [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+       <width>120</width>
+       <height>0</height>
+      </size>
+     </property>
+    </widget>
+   </item>
+   <item row="1" column="0">
+    <widget class="QLabel" name="lblShininess">
+     <property name="toolTip">
+      <string>How shiny smooth surfaces are.</string>
```
Comment (wonder-sk) [inline src/ui/3d/metalroughmaterialwidget.ui:46]: the tooltips for base color, metalness and roughness probably need updating

**Review by wonder-sk — APPROVED** [review]
Awesome, thanks for adding PBR material! :clap: 

By the way, have you had any luck understanding tooling to deal with the JSON shader graphs? The whole thing with QShaderProgramBuilder looks quite opaque to me, and I am already dreading the moment when we would need to adjust that shader graph for some reason...

File: src/ui/3d/metalroughmaterialwidget.ui (hunk 1 above)
Comment (nyalldawson) [reply src/ui/3d/metalroughmaterialwidget.ui:46]: >  The whole thing with QShaderProgramBuilder looks quite opaque to me, and I am already dreading the moment when we would need to adjust that shader graph for some reason...

100% agree! My gut feeling is that we'll probably need to replace the shader with a different (non-qt? a copy of quick3d's?) shader.

**Comment by wonder-sk** [comment]
@nyalldawson I have had a play with the metal roughness material, and found out that the newly added `src/3d/shaders/metalrough.frag` file actually does not get used at all :grimacing:  The whole shader program builder is a bit more complicated, in addition to the graph with nodes/edges, there is also a builtin library of node "prototypes" in qt3d, where each prototype defines inputs/outputs of a node, and how such node transcribes to actual shader code (in different graphics APIs) - [here is metalRoughFunction definition](https://github.com/qt/qt3d/blob/afc99d91af4d27011c777aac101593bf1e1be04c/src/render/materialsystem/prototypes/default.json#L633), with an include of `:/shaders/gl3/metalrough.inc.frag` from within qt3d resources.

Apparently the shader graph JSON can also contain new prototypes in addition to nodes/edges (see [QShaderGraphLoader::load()](https://github.com/qt/qt3d/blob/dev/src/render/shadergraph/qshadergraphloader.cpp#L62)), so in theory we could add our own variant of metalRoughFunction as a prototype and use it in the shader graph JSON. But with the lack of tooling, that seems like a future source of headaches.

My understanding is that the qt3d shader builders are trying to accomplish two things at once:

1. support multiple graphics APIs (GL, GL ES, Qt's RHI),
2. support conditional build of shader code - e.g. baseColor/metalness/roughness can be either a single value given in a uniform, or it can be sampled from a texture - and the shader builder will build the right code, without having to keep duplicate shader code for all the permutations

In our use case where we only support OpenGL, so things are simpler, but we would still want conditional build of shader code. The good news is that GLSL language [supports all the usual C preprocessor macros](https://registry.khronos.org/OpenGL/specs/es/3.2/GLSL_ES_Specification_3.20.html#preprocessor) apart from `#include` which is handled by frameworks like qt3d separately.

So for metal/rough material, I am wondering whether not to take the route similar to what three.js is doing, where each material has shader code with lots of includes ([meshbasic.glsl.js](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderLib/meshbasic.glsl.js)) and then there is a shader chunk library with pieces of shader code with #ifdefs ([color_vertex.glsl.js](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/color_vertex.glsl.js)).

An alternative would be to disable all the functionality we do not currently use (such as baseColor, metalness, roughness as textures, normal texture, ambient occlusion texture), keep just the basic shader code, and only start adding complexity with conditional building of the shader once needed (we will probably want baseColor as a texture at some point, but I doubt the other bits would be needed anytime soon).

Keen to hear your thoughts :wink:

**Pull request merged**